protocol/ibd/selected_tip.go

```diff
@@ -0,0 +1,133 @@
+package ibd
+
+import (
+	"errors"
+	"github.com/kaspanet/kaspad/blockdag"
+	"github.com/kaspanet/kaspad/netadapter/router"
+	"github.com/kaspanet/kaspad/protocol/common"
+	peerpkg "github.com/kaspanet/kaspad/protocol/peer"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"time"
+)
+
+const minDurationToRequestSelectedTips = time.Minute
+
+func requestSelectedTipsIfRequired(dag *blockdag.BlockDAG) error {
+	if hasRecentlyReceivedBlock(dag) {
+		return nil
+	}
+	return requestSelectedTips(dag)
+}
+
+func hasRecentlyReceivedBlock(dag *blockdag.BlockDAG) bool {
+	return dag.Now().Sub(dag.SelectedTipHeader().Timestamp) > minDurationToRequestSelectedTips
+}
+
+func requestSelectedTips(dag *blockdag.BlockDAG) error {
+	for _, peer := range peerpkg.ReadyPeers() {
+		peer.RequestSelectedTipIfRequired()
+	}
+	return nil
+}
+
+// RequestSelectedTip waits for selected tip requests and handles them
+func RequestSelectedTip(incomingRoute *router.Route,
+	outgoingRoute *router.Route, peer *peerpkg.Peer, dag *blockdag.BlockDAG) error {
+	for {
+		peer.WaitForSelectedTipRequests()
+
+		// We run the flow inside a func so that the defer is called at its end
+		err := func() error {
+			defer peer.FinishRequestingSelectedTip()
+
+			shouldContinue, err := requestSelectedTip(outgoingRoute)
+			if err != nil {
+				return err
+			}
+			if !shouldContinue {
+				return nil
+			}
+
+			peerSelectedTipHash, shouldContinue, err := receiveSelectedTip(incomingRoute)
+			if err != nil {
+				return err
+			}
+			if !shouldContinue {
+				return nil
+			}
+			err = peer.SetSelectedTipHash(peerSelectedTipHash)
+
+			return StartIBDIfRequired(dag)
+		}()
+		if err != nil {
+			return err
+		}
+	}
+}
+
+func requestSelectedTip(outgoingRoute *router.Route) (shouldContinue bool, err error) {
+	msgGetSelectedTip := wire.NewMsgGetSelectedTip()
+	isOpen, err := outgoingRoute.EnqueueWithTimeout(msgGetSelectedTip, common.DefaultTimeout)
```

Comment:
You can `return outgoingRoute.EnqueueWithTimeout(//...)` at this point, will be equivalent 
(assuming you are not combining requestSelectedTip and receiveSelectedTip as requested in previous comment)